src/QsCompiler/QirGeneration/QIR/Types.cs

```diff
@@ -142,6 +159,28 @@ internal Types(Context context, Func<QsQualifiedName, QsCustomType?> getTypeDecl
 
         // internal helpers to simplify common code
 
+        /// <summary>
+        /// Creates a struct type with the given name and element types if it does not already exist.
+        /// If a struct with the given name already exists, the existing type is returned without validating the element types.
+        /// </summary>
+        internal IStructType NamedStruct(string name, params ITypeRef[] elementTypes)
+        {
+            if (!this.namedStructs.TryGetValue(name, out var type))
+            {
+                type = this.context.CreateStructType(name, false, elementTypes);
+                this.namedStructs.Add(name, type);
+            }
+
+            return type;
+        }
+
+        /// <summary>
+        /// Creates the LLVM data type used for stack allocated arrays whose size will be resolved
+        /// as part of a compilation step and only after the initial QIR generation.
+        /// </summary>
+        internal IStructType NamedLlvmArray(string storageName, ITypeRef elementType) =>
+            Emission.ArrayValue.StackAllocatedArrayType(this, elementType, 0u, storageName);
```

Comment:
(nit) Passing a hard-coded value (magic number `0u`) is not very informative. For my code in such cases I prefer to add the param name:
```suggestion
            Emission.ArrayValue.StackAllocatedArrayType(this, elementType, nrElements: 0u, storageName);
```